hips/hip-9999.md

```diff
@@ -0,0 +1,83 @@
+---
+hip: 9999
+title: "Enhanced logging library for Helm CLI"
+authors: [ "Evans Mungai <mbuevans@gmail.com>" ]
+created: "2024-11-22"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+This proposal introduces an improved logging library for Helm CLI to enhance logging capabilities for troubleshooting, development, and debugging. By using newer logging features such as log formatting, different log levels and verbosity levels, developers and CLI users gain flexibility when deciding how much logging information they want to consume, and in what format.
+
+## Motivation
+
+The current logging mechanism in Helm CLI lacks granularity and structure. Users and developers face difficulties when:
+
+- Troubleshooting Helm CLI commands due to inability to increase the details that can get logged. More logs come in handy when submitting bug reports for example.
+- Instrumenting code with clear and consistent logging conventions for contributors, as well as consumers of CLI logs.
+
+The proposed solution addresses these gaps by integrating a logging library that supports log levels, verbosity, and structured output.
+
+## Rationale
+
+Two libraries were evaluated:
+
+1. **`slog`:**
+   - Provides log levels (`info`, `debug`, `error`, `warning`).
+   - Offers structured logging.
+   - Lightweight and straightforward to integrate.
+   - Is part of the standard library
+
+2. **`klog`:**
+   - Provides log levels and also verbosity levels, enabling more granular control. Verbosity level controls the amount of detail in logs, enabling users to filter messages based on their importance or relevance.
+   - Well-aligned with Kubernetes development standards.
+   - Allows getting instrumented Kubernetes client library.
+
+While both libraries meet the basic requirements, `klog` offers richer features for developers familiar with Kubernetes' logging patterns, making it the preferred choice for this proposal.
```

Comment:
There are two parts to Helm. There is the SDK and then the client that uses the SDK.

The SDK shouldn't use an implementation of logging. It should accept a logger from any application that uses it and follows the right interface. So, the SDK should use either `slog` or `logr`. It may have a default one but that needs to be easy to override.

Note, SDK users may not be developers familiar with Kubernetes. There will be people familiar with and those not familiar with it.

Then there is the Helm client (a.k.a CLI). This should pick an implementation. It could be slog, klog, zap, or something else. Note, klog appears to support the slog interface ([see this example](https://github.com/kubernetes/klog/blob/main/contextual_slog_example_test.go)).

We need to make decisions on both of these. If we go the klog route for the Helm Client we should document the log levels to use and what they mean. There is no common clarity on what levels to use for what. This is one of the weakness of this model. What level should you set for what purposes?